addons/crm/models/crm_team.py

```diff
@@ -113,12 +146,329 @@ def _alias_get_creation_values(self):
+
+        # log a note in case of manual assign
+        log_action = _("Lead Assignment done by %(user_name)s", user_name=self.env.user.name)
+        log_message = "<p>%s<br /><br />%s</p>" % (log_action, html_message)
+        self._message_log_batch(bodies=dict((team.id, log_message) for team in self))
+
+        return {
+            'type': 'ir.actions.client',
+            'tag': 'display_notification',
+            'params': {
+                'type': 'success',
+                'title': _("Leads Assigned"),
+                'message': notif_message,
+                'next': {
+                    'type': 'ir.actions.act_window_close'
+                },
+            }
+        }
+
+    def _action_assign_leads(self, work_days=2):
+        """ Private method for lead assignment. This method both
+
+          * assigns leads to teams given by self;
+          * assigns leads to salespersons belonging to self;
+
+        See sub methods for more details about assign process.
+
+        :param int work_days: see ``CrmTeam.action_assign_leads()``;
+
+        :return teams_data, members_data: structure-based result of assignment
+          process. For more details about data see ``CrmTeam._assign_leads()``
+          and ``CrmTeamMember._assign_and_convert_leads``;
+        """
+        if not self.env.user.has_group('sales_team.group_sale_manager') and not self.env.user.has_group('base.group_system'):
+            raise exceptions.UserError(_('Lead/Opportunities automatic assignment is limited to managers or administrators'))
+
+        _logger.info('### START Lead Assignment (%d teams, %d sales persons, %d work_days)' % (len(self), len(self.crm_team_member_ids), work_days))
+        teams_data = self._allocate_leads()
+        _logger.info('### Team repartition done. Starting salesmen assignment.')
+        members_data = self.crm_team_member_ids._assign_and_convert_leads(work_days=work_days)
+        for member_id, member_info in members_data.items():
+            _logger.info('-> member %s: assigned %s' % (member_id, member_info["assign"]))
+        _logger.info('### END Lead Assignment')
+        return teams_data, members_data
+
+    def _action_assign_leads_log(self, teams_data, members_data, html=False):
+        """ Tool method to prepare notification about assignment process result.
+
+        :param teams_data: see ``CrmTeam._allocate_leads()``;
+        :param members_data: see ``CrmTeamMember._assign_and_convert_leads()``;
+        :param html: if True target formatted content in html (suitable for a
+          log on a document); else target plaintext content (suitable for a
+          small notification).
+
+        :return string: formatted log
+        """
+        # extract some statistics
+        assigned = sum(len(teams_data[team]['assign']) + len(teams_data[team]['merged']) for team in self)
+        duplicates = sum(len(teams_data[team]['duplicates']) for team in self)
+        members = len(members_data.keys())
```

Comment:
If members_data is a dict, len(dict) should return the correct result. No need to fetch keys.